From 7f498cf649827068616ee57958e60195ac2ee68c Mon Sep 17 00:00:00 2001
From: Jiajian Wu <jair.wu@rock-chips.com>
Date: Fri, 19 Nov 2021 10:38:01 +0800
Subject: [PATCH 92/94] drm/bridge: synopsys: Add hook_plugged_cb for
 dw-hdmi-qp reporting connector status

Signed-off-by: Jiajian Wu <jair.wu@rock-chips.com>
Change-Id: I3103df9f379dfe98a4eb54dc5d3ce9f407038d8c
Signed-off-by: Detlev Casanova <detlev.casanova@collabora.com>

Gbp-Pq: Name 0173-drm-bridge-synopsys-Add-hook_plugged_cb-for-dw-hdmi-.patch
---
 .../drm/bridge/synopsys/dw-hdmi-qp-i2s-audio.c   | 11 +++++++++++
 drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c     | 16 ++++++++++++++++
 include/drm/bridge/dw_hdmi.h                     |  3 +++
 3 files changed, 30 insertions(+)

diff --git a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp-i2s-audio.c b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp-i2s-audio.c
index 43a21dd54..e07c2b423 100644
--- a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp-i2s-audio.c
+++ b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp-i2s-audio.c
@@ -116,12 +116,23 @@ static int dw_hdmi_qp_i2s_get_dai_id(struct snd_soc_component *component,
 	return -EINVAL;
 }
 
+static int dw_hdmi_qp_i2s_hook_plugged_cb(struct device *dev, void *data,
+					  hdmi_codec_plugged_cb fn,
+					  struct device *codec_dev)
+{
+	struct dw_hdmi_qp_i2s_audio_data *audio = data;
+	struct dw_hdmi_qp *hdmi = audio->hdmi;
+
+	return dw_hdmi_qp_set_plugged_cb(hdmi, fn, codec_dev);
+}
+
 static struct hdmi_codec_ops dw_hdmi_qp_i2s_ops = {
 	.hw_params	= dw_hdmi_qp_i2s_hw_params,
 	.audio_startup  = dw_hdmi_qp_i2s_audio_startup,
 	.audio_shutdown	= dw_hdmi_qp_i2s_audio_shutdown,
 	.get_eld	= dw_hdmi_qp_i2s_get_eld,
 	.get_dai_id	= dw_hdmi_qp_i2s_get_dai_id,
+	.hook_plugged_cb = dw_hdmi_qp_i2s_hook_plugged_cb,
 };
 
 static int snd_dw_hdmi_qp_probe(struct platform_device *pdev)
diff --git a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c
index e31262b39..fcc785e2f 100644
--- a/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c
+++ b/drivers/gpu/drm/bridge/synopsys/dw-hdmi-qp.c
@@ -161,6 +161,22 @@ struct dw_hdmi_qp {
 	struct regmap *regm;
 };
 
+int dw_hdmi_qp_set_plugged_cb(struct dw_hdmi_qp *hdmi, hdmi_codec_plugged_cb fn,
+			      struct device *codec_dev)
+{
+	bool plugged;
+
+	mutex_lock(&hdmi->mutex);
+	hdmi->plugged_cb = fn;
+	hdmi->codec_dev = codec_dev;
+	plugged = hdmi->last_connector_result == connector_status_connected;
+	handle_plugged_change(hdmi, plugged);
+	mutex_unlock(&hdmi->mutex);
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(dw_hdmi_qp_set_plugged_cb);
+
 static void hdmi_set_cts_n(struct dw_hdmi_qp *hdmi, unsigned int cts,
 			   unsigned int n)
 {
diff --git a/include/drm/bridge/dw_hdmi.h b/include/drm/bridge/dw_hdmi.h
index 67b3671c9..c66054087 100644
--- a/include/drm/bridge/dw_hdmi.h
+++ b/include/drm/bridge/dw_hdmi.h
@@ -240,5 +240,8 @@ void dw_hdmi_qp_set_audio_infoframe(struct dw_hdmi_qp *hdmi,
 				    struct hdmi_codec_params *hparms);
 void dw_hdmi_qp_audio_enable(struct dw_hdmi_qp *hdmi);
 void dw_hdmi_qp_audio_disable(struct dw_hdmi_qp *hdmi);
+bool dw_hdmi_qp_connected(struct dw_hdmi_qp *hdmi);
+int dw_hdmi_qp_set_plugged_cb(struct dw_hdmi_qp *hdmi, hdmi_codec_plugged_cb fn,
+			      struct device *codec_dev);
 
 #endif /* __IMX_HDMI_H__ */
-- 
2.46.0

